packages/hardware-ledger/src/LedgerKeyAgent.ts

```diff
@@ -148,98 +165,207 @@ export class LedgerKeyAgent extends KeyAgentBase {
     this.#communicationType = serializableData.communicationType;
   }
 
-  static findKeyAgentByCommunicationTypeAndDevicePath(communicationType: CommunicationType, devicePath?: string) {
-    return this.deviceConnections?.find(
-      (connection) => connection.communicationType === communicationType && connection.devicePath === devicePath
+  private static async findConnectionByCommunicationTypeAndDevicePath(
+    communicationType: CommunicationType,
+    nodeHidDevicePath?: string,
+    device?: LedgerDevice
+  ): Promise<LedgerConnection | null> {
+    const matchingConnectionData = this.deviceConnections?.find(
+      (connection) =>
+        connection.communicationType === communicationType &&
+        connection.nodeHidDevicePath === nodeHidDevicePath &&
+        connection.device === device
     );
+    if (!matchingConnectionData) return null;
+
+    try {
+      await this.testConnection(matchingConnectionData.deviceConnection);
+    } catch (error) {
+      if (error instanceof errors.TransportError && error.message.includes(CARDANO_APP_CONNECTION_ERROR_MESSAGE)) {
+        this.deviceConnections = this.deviceConnections.filter(
+          (connectionData) => connectionData !== matchingConnectionData
+        );
+        return null;
+      }
+
+      throw error;
+    }
+
+    return matchingConnectionData.deviceConnection;
   }
 
   /**
    * @throws TransportError
    */
-  static async getHidDeviceList(communicationType: CommunicationType): Promise<string[]> {
+  private static async getHidDeviceList(communicationType: CommunicationType): Promise<string[]> {
     try {
-      return communicationType === CommunicationType.Node ? TransportNodeHid.list() : TransportWebHID.list();
+      return communicationType === CommunicationType.Node ? TransportNodeHid.list() : TransportWebUSB.list();
     } catch (error) {
       throw new errors.TransportError('Cannot fetch device list', error);
     }
   }
 
+  private static async closeTransport(transport: LedgerTransportType) {
+    try {
+      await transport.close();
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error('Failed to close the transport', transport, error);
```

Comment:
Please add `Logger` as a dependency instead
```suggestion
      this.#logger.error('Failed to close the transport', transport, error);
```

Reply:
Oh I see, didn't notice it's in a static method. In that case it wouldn't be as clean to inject. If this log isn't essential (seems like not?) then I suggest replacing this line with a comment about the implications of failing to close (if any)

Reply:
I decided to step back and remove this method. It is a low probability that the `transport.close` will fail and affect the logic, so IMO it is not worth making additional effort. If it fails we will notice that as a console error and that's enough.